Gems/Atom/RPI/Code/Source/RPI.Public/Pass/Pass.cpp

```diff
@@ -1426,6 +1433,57 @@ namespace AZ
             m_flags.m_pipelineStatisticsQueryEnabled = enable;
         }
 
+        bool Pass::ReadbackAttachment(AZStd::shared_ptr<AttachmentReadback> readback, const Name& slotName, bool useOutput)
+        {
+            // Return false if it's already readback
+            if (m_attachmentReadback)
+            {
+                AZ_Warning("Pass", false, "ReadbackAttachment: skip readback pass [%s] slot [%s]because there is an another active readback", m_name.GetCStr(), slotName.GetCStr());
+                return false;
+            }
+            uint32_t bindingIndex = 0;
+            for (auto& binding : m_attachmentBindings)
+            {
+                if (slotName == binding.m_name)
+                {
+                    RHI::AttachmentType type = binding.m_attachment->GetAttachmentType();
+                    if (type == RHI::AttachmentType::Buffer || type == RHI::AttachmentType::Image)
+                    {
+                        RHI::AttachmentId attachmentId = binding.m_attachment->GetAttachmentId();
+
+                        // Append slot index and pass name so the read back's name won't be same as the attachment used in other passes.
+                        AZStd::string readbackName = AZStd::string::format("%s_%d_%s", attachmentId.GetCStr(),
+                            bindingIndex, GetName().GetCStr());
+                        if (readback->ReadPassAttachment(binding.m_attachment.get(), AZ::Name(readbackName)))
+                        {
+                            m_readbackOutput = true;
+                            // The useOutput is only meaningful if the attachment is used for InputOutput.
+                            if (binding.m_slotType == PassSlotType::InputOutput)
+                            {
+                                m_readbackOutput = useOutput;
+                            }
+                            m_attachmentReadback = readback;
+                            return true;
+                        }
+                        return false;
+                    }
+                }
+                bindingIndex++;
+            }
+            AZ_Warning("Pass", false, "ReadbackAttachment: failed to find slot [%s] from pass [%s]", slotName.GetCStr(), m_name.GetCStr());
+            return false;
+        }
+
+        void Pass::UpdateReadbackAttachment(FramePrepareParams params, bool beforeAddScopes)
+        {
+            if (beforeAddScopes == !m_readbackOutput && m_attachmentReadback)
+            {
+                // Read the attachment for one frame. The reference can be released afterwards
+                m_attachmentReadback->FrameBegin(params);
+                m_attachmentReadback = nullptr;
```

Comment:
nit: Calling Reset() makes it clearer.

Reply:
It can't call the Reset since this AttachmentReadback is still alive. Here it's just reduce the reference.

Reply:
Reset() will just release the current shared_ptr, not force a deletion.

Reply:
ah okay. AttachmentReadback class has a reset function too.